cmd/gen-ca/main.go

```diff
@@ -0,0 +1,378 @@
+			CommonName:   profile.Subject.CommonName,
+			Organization: []string{profile.Subject.Organization},
+			Country:      []string{profile.Subject.Country},
+		},
+		NotBefore:             notBefore,
+		NotAfter:              notAfter,
+		OCSPServer:            []string{profile.OCSPURL},
+		CRLDistributionPoints: []string{profile.CRLURL},
+		IssuingCertificateURL: []string{profile.IssuerURL},
+		PolicyIdentifiers:     policyOIDs,
+		KeyUsage:              x509.KeyUsageCertSign & x509.KeyUsageCRLSign,
+		SubjectKeyId:          keyHash[:],
+	}
+
+	return cert, nil
+}
+
+func main() {
+	module := flag.String("module", "", "PKCS#11 module to use")
+	slot := flag.Uint("slot", 0, "Slot signing key is in")
+	pin := flag.String("pin", "", "PIN for slot if not using PED to login")
+	label := flag.String("label", "", "Signing key label")
+	id := flag.String("id", "", "Signing key ID hexadecimal (simplified format, i.e. ffff")
+	profilePath := flag.String("profile", "", "Path to certificate profile")
+	pubKeyPath := flag.String("publicKey", "", "Path to public key for certificate") // this could also be in the profile
+	issuerPath := flag.String("issuer", "", "Path to issuer cert if generating an intermediate")
+	outputPath := flag.String("output", "", "Path to store generated PEM certificate")
+	flag.Parse()
+
+	if *module == "" {
+		log.Fatal("--module is required")
+	}
+	if *label == "" {
+		log.Fatal("--label is required")
+	}
+	if *id == "" {
+		log.Fatal("--id is required")
+	}
+	if *profilePath == "" {
+		log.Fatal("--profile is required")
+	}
+	if *pubKeyPath == "" {
+		log.Fatal("--publicKey is required")
+	}
+	if *outputPath == "" {
+		log.Fatal("--output is required")
+	}
+
+	ctx, session, err := pkcs11helpers.Initialize(*module, *slot, *pin)
+	if err != nil {
+		log.Fatalf("Failed to setup session and PKCS#11 context: %s", err)
+	}
+	log.Println("Opened PKCS#11 session")
+
+	privKey, err := getKey(ctx, session, *label, *id)
+	if err != nil {
+		log.Fatalf("Failed to retrieve private key handle: %s", err)
+	}
+	log.Println("Retrieved private key handle")
+
+	profileBytes, err := ioutil.ReadFile(*profilePath)
+	if err != nil {
+		log.Fatalf("Failed to read certificate profile %q: %s", *profilePath, err)
+	}
+	var profile certProfile
+	err = json.Unmarshal(profileBytes, &profile)
+	if err != nil {
+		log.Fatalf("Failed to parse certificate profile: %s", err)
+	}
+
+	pubPEMBytes, err := ioutil.ReadFile(*pubKeyPath)
+	if err != nil {
+		log.Fatalf("Failed to read public key %q: %s", *pubKeyPath, err)
+	}
+	pubPEM, _ := pem.Decode(pubPEMBytes)
+	if pubPEM == nil {
+		log.Fatal("Failed to parse public key")
+	}
+	pub, err := x509.ParsePKIXPublicKey(pubPEM.Bytes)
+	if err != nil {
+		log.Fatalf("Failed to parse public key: %s", err)
+	}
+
+	certTmpl, err := constructCert(ctx, &profile, pubPEM.Bytes, session)
+	if err != nil {
+		log.Fatalf("Failed to construct certificate from profile: %s", err)
+	}
+	log.Println("Generated tbs certificate")
```

Comment:
*"Generated certificate template." This isn't a `tbsCertificate` because it's not serialized DER.